### ISSUE TITLE
Unlinking a file yields "could not locate artifact"

### ISSUE THREAD

**Issue by mentaal** [devopshq/artifactory#36]
I tried unlinking a file from artifactory using `dohq_artifactory-0.4.102`
I got an error 404 "Could not locate artifact"

I then tried uninstalling the aformentioned version of dohq-artifactory and installing `artifactory-0.1.17` which seemed to successfully unlink the file.

**Comment by mentaal** [comment]
Below is the stack trace, with some names suitably altered:

    RuntimeError                              Traceback (most recent call last)
    <ipython-input-24-57003bea8697> in <module>()
    ----> 1 p.unlink()

    c:\Python37-32\lib\pathlib.py in unlink(self)
       1264         if self._closed:
       1265             self._raise_closed()
    -> 1266         self._accessor.unlink(self)
       1267
       1268     def rmdir(self):

    c:\users\a_user\python_envs\a_proj\lib\site-packages\artifactory.py in unlink(self, pathobj)
        613
        614         if code not in [200, 202, 204]:
    --> 615             raise RuntimeError("Failed to delete file: %d '%s'" % (code, text))
        616
        617     def touch(self, pathobj):

    RuntimeError: Failed to delete file: 404 '{
      "errors" : [ {
        "status" : 404,
        "message" : "Could not locate artifact 'blah-blahblah:some_mod/some_mod-1.8.20.zip'."
      } ]
    }'

**Comment by mentaal** [comment]
Actually this may be been something to do without providing the `auth` parameter as I get the same error with the `artifactory` library when I forgot to provide it.

**Comment by mentaal** [comment]
This happened again to me today and I was certain that I had added the relevant authentication token. 
This time, I tried to upload a file and got the same "Unauthorized" message. 

I uninstalled dohq_artifactory-0.4.103, reinstalled artifactory and the upload worked just fine. 

Could this be a compatibility issue with the version of artifactory that we have in house?

**Comment by beliaev-maksim** [comment]
also, here could be an issue if colon was used. #254 might fixed it

@mentaal 
can you check on the latest version?

**Comment by mentaal** [comment]
Hi @beliaev-maksim ,
Thanks for the prompt and apologies for the delay in responding. 
I did try deleting a file in my company's artifactory instance using `unlink` but I end up with a `404`.

I'm essentially doing this:

    the_dir = artifactory_path / 'some_directory'

    for f in the_dir.iterdir():
        print(f'removing: {f}...')
        f.unlink()

I end up with the error:

    FileNotFoundError: Failed to delete file: 404 '{\n  "errors" : [ {\n    "status" : 404,\n    "message" : "Could not locate artifact \'the_repo:some_directory/the_file\'."\n  } ]\n}'

I definitely have the correct path because I can iterate through the files and print their paths.

Though this could very well be something on my end. Maybe permissions or something? Anyway, it doesn't seem like I'm in a good position to validate this change now unfortunately.

    $ pip show dohq-artifactory
    Name: dohq-artifactory
    Version: 0.7.658
    Summary: A Python interface to Artifactory
    Home-page: https://devopshq.github.io/artifactory/
    Author: Alexey Burov
    Author-email: aburov@ptsecurity.com
    License: MIT License
    Location: c:\users\gkuhn\python_envs\apollo\lib\site-packages
    Requires: requests, PyJWT, python-dateutil
    Required-by:

**Comment by beliaev-maksim** [comment]
@mentaal 
should not be a permission issue, I just validated and it returns 403 as it should be when I try to unlink file without permissions
do you see that issue for each file in artifactory?

do you have permissions to upload (deploy) a file? Can you try to upload the file, print its name and then unlink?

**Comment by mentaal** [comment]
@beliaev-maksim ok I tried this:

    with TemporaryDirectory() as td:
        tf = Path(td) / 'temp_file.txt'
        tf.write_text('testing testing, is this thing on?...')

        firmware_dir.deploy_file(tf)


    for f in firmware_dir.iterdir():
        print(f'found file: {f}')

    to_remove = firmware_dir / 'temp_file.txt'

    if to_remove.exists():
        print(f'found file: {to_remove}. Removing...')
        to_remove.unlink()



and got this result: (I'm replaced file names and urls etc)

    $ python proj_artifactory.py
    found file: https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
    found file: https://artifactory.company.com/artifactory/company-nuget/proj/test_pkg_dont_use.nupkg
    found file: https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt. Removing...
    Traceback (most recent call last):
      File "proj_artifactory.py", line 39, in <module>
        to_remove.unlink()
      File "C:\python38-32\lib\pathlib.py", line 1321, in unlink
        self._accessor.unlink(self)
      File "C:\Users\gkuhn\python_envs\proj\lib\site-packages\artifactory.py", line 898, in unlink
        raise FileNotFoundError("Failed to delete file: {} {!r}".format(code, text))
    FileNotFoundError: Failed to delete file: 404 '{\n  "errors" : [ {\n    "status" : 404,\n    "message" : "Could not locate artifact \'company-nuget:proj/temp_file.txt\'."\n  } ]\n}'

**Comment by mentaal** [comment]
Sorry, just saw your comment about special characters. the only non alpha-numeric character in the repo name is a `-`. 
The repo name is of the form: `<prefix-nuget`>

**Comment by beliaev-maksim** [comment]
@mentaal 
can you please try develop version, install it using command below, run same script as you shown, send output

`pip install git+https://github.com/beliaev-maksim/artifactory.git@test_unlink`

**Comment by mentaal** [comment]
Hi @beliaev-maksim,

Ok I've updated to the test_unlink path as you requested and reran:

    $ python proj_artifactory.py
    https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
    https://artifactory.company.com/artifactory/company-nuget/proj/test_pkg_dont_use.nupkg
    found file: https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
    found file: https://artifactory.company.com/artifactory/company-nuget/proj/test_pkg_dont_use.nupkg
    found file: https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt. Removing...
    unqouted https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
    quoted https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
    Traceback (most recent call last):
      File "proj_artifactory.py", line 39, in <module>
        to_remove.unlink()
      File "C:\python38-32\lib\pathlib.py", line 1321, in unlink
        self._accessor.unlink(self)
      File "C:\Users\gkuhn\python_envs\proj\lib\site-packages\artifactory.py", line 902, in unlink
        self.rest_del(
      File "C:\Users\gkuhn\python_envs\proj\lib\site-packages\artifactory.py", line 718, in rest_del
        response.raise_for_status()
      File "C:\Users\gkuhn\python_envs\proj\lib\site-packages\requests\models.py", line 943, in raise_for_status
        raise HTTPError(http_error_msg, response=self)
    requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt

**Comment by beliaev-maksim** [comment]
@mentaal 
thanks,
at least now error message is consistent and we see correct URL is used

can you please do another check:
1. create file **manually** on Artifactory
2. then run following snippet (optionally adding auth if required)
~~~python
from artifactory import ArtifactoryPath

path = ArtifactoryPath(
    "https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt"
)

if path.exists():
    print("exists")
    path.unlink()
~~~

here we can check if children of the object have some issue. Also, we can check if file has stats

**Comment by mentaal** [comment]
As an aside, do you know why this happens?

    urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionAbortedError(10053, 'An established connection was aborted by the software in your host machine', None, 10053, None))

I occasionally see this when I run it from my terminal. 

Anyway, I manually created a new file: `temp_file.txt` this time and uploaded it using the artifactory web UI. 
I ran the script you suggested and got:

exists
unlinking...
unqouted https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
quoted https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt
Traceback (most recent call last):
  File "apollo_artifactory.py", line 54, in <module>
    path.unlink()
  File "C:\python38-32\lib\pathlib.py", line 1321, in unlink
    self._accessor.unlink(self)
  File "C:\Users\gkuhn\python_envs\apollo\lib\site-packages\artifactory.py", line 902, in unlink
    self.rest_del(
  File "C:\Users\gkuhn\python_envs\apollo\lib\site-packages\artifactory.py", line 718, in rest_del
    response.raise_for_status()
  File "C:\Users\gkuhn\python_envs\apollo\lib\site-packages\requests\models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://artifactory.company.com/artifactory/company-nuget/proj/temp_file.txt

The URL looks fine too. If you can't reproduce something like this, surely this must suggest it's something on my end or my artifactory instance's environment or configuration?

**Comment by beliaev-maksim** [comment]
@mentaal 
it is not really related to the library, you can google more for "protocol error 10053"

btw, 
can you download file instead of unlinking it?
I am curious if that will also produce an error

**Comment by mentaal** [comment]
Thanks @beliaev-maksim the permissions that is discussed in those links must be it. I didn't have required permissions in the repo I was trying to delete from. In fact I had so little permissions, I couldn't even tell! It was only when I explored a different repo where I did have more permissions, the delete options showed up in the web UI. 

Using `unlink` in this other repo works fine. 

Thanks for all your help in tracking this down. I think this can be closed.
